### DISCUSSION TITLE
Mechanical power in cost function

### DISCUSSION THREAD

**Discussion by SmbdOnceTldMe** [Simple-Robotics/aligator#251] [Q&A]
Hello!
Thanks fo such great library!

Could you give me some hints how I can add a mechanical power to cost function? I need it for biped locomotion problem
Do I understand correctly that I should inherit  `aligator.CostAbstract` (like you did [here](https://github.com/Simple-Robotics/aligator/blob/ca559224f61a693c19a383a7bb4671007e368305/tests/python/test_example_problem.py#L72)) to create my own cost class? Here you use `QuadraticDistanceCost` class as a 'basis' to guarantee desired state. Which analogue of  `QuadraticDistanceCost` I can use to minimize `u.T*x[nq:]` value?

**Comment by ManifoldFR** [comment]
Hello, thanks for your interest in aligator!

If your problem is Euclidean, you can just use the `QuadraticCost` class and set the cross-term matrix appropriately so that an `u^\top v` term can appear.
Something like

$$
  N = \begin{bmatrix} 0_{n_u\times n_v} & I_{n_u\times n_v} \end{bmatrix}
$$

Otherwise, indeed a subclass to `aligator.CostAbstract` would work.

**Reply by SmbdOnceTldMe** [reply]
Thanks for your reply!
Yes, I tried this but I got following error
![photo_2024-12-05_15-20-16](https://github.com/user-attachments/assets/88460438-643d-4fa5-8d93-ae0551d19810)
As I understand because I have a space with unequal dimensions of `space.ndx = space.nx - 1` 
(I checked it with Talos model as well and got similar error)

**Comment by ManifoldFR** [comment]
Oh if it's Talos then it's non-Euclidean ! You'll have to write your own
class in that case.

**Reply by SmbdOnceTldMe** [reply]
Ok, thanks a lot
I'll try it
